### PR TITLE
Dark Steel Work

### PR DESCRIPTION
**Pull request by CrazyPants**
# Description

Added the Direct upgrade
Added the dual energy/durability bar
Added recipes for tools and upgrades

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: enderio-base/src/main/java/com/enderio/base/client/renderer/DarkSteelDurabilityRenderer.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        RenderSystem.disableTexture();
+        RenderSystem.disableBlend();
+        Tesselator tesselator = Tesselator.getInstance();
+        BufferBuilder bufferbuilder = tesselator.getBuilder();
+
+        int offset = 12;
+        int barWidth = Math.round(13.0F - (float)getFilledRatio(pStack) * 13.0F);
+        fillRect(bufferbuilder, pXPosition + 2, pYPosition + offset, 13, 1, 0, 0, 0);
+        fillRect(bufferbuilder, pXPosition + 2, pYPosition + offset, barWidth, 1,
+            ENERGY_BAR_RGB >> 16 & 255, ENERGY_BAR_RGB >> 8 & 255, ENERGY_BAR_RGB & 255);
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/client/renderer/DarkSteelDurabilityRenderer.java]: Maybe use FastColor.ARGB32 for better readability

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java [hunk 2]
```diff
@@ -41,7 +48,41 @@ public float getDestroySpeed(ItemStack pStack, BlockState pState) {
 
     @Override
     public boolean mineBlock(ItemStack pStack, Level pLevel, BlockState pState, BlockPos pPos, LivingEntity pEntityLiving) {
-        //TODO: Entire tree chopping
+        if(pEntityLiving instanceof Player player) {
+            if(pEntityLiving.isCrouching() && pState.is(BlockTags.LOGS) && EnergyUtil.getEnergyStored(pStack) > 0) {
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java]: spaces between if and (

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java [hunk 3]
```diff
@@ -71,6 +112,48 @@ private boolean hasFork(ItemStack stack) {
         return DarkSteelUpgradeable.hasUpgrade(stack, ForkUpgrade.NAME);
     }
 
+    private void fellTree(Level level, BlockPos pos, Set<BlockPos> checkedPos, Set<BlockPos> toChop, int maxBlocks, Block targetBock) {
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java]: could u add javadoc for each parameter?

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java [hunk 4]
```diff
@@ -71,6 +112,48 @@ private boolean hasFork(ItemStack stack) {
...
+    private void surrounding(Set<BlockPos> res, BlockPos pos) {
+        for(Direction dir : Direction.values()) {
+            if(dir != DOWN && dir != UP) {
+                res.add(pos.relative(dir));
+            }
+        }
+        res.add(pos.north().east());
+        res.add(pos.north().west());
+        res.add(pos.south().east());
+        res.add(pos.south().west());
```
Comment (agnor99) [inline enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java:143]: suggestion:
I think this is more readable
```suggestion
        for (int x = -1; x <= 1; x++) {
            for (int z = -1; z <= 1; z++) {
                if (x != 0 || z != 0) {
                  res.add(pos.offset(x, 0, z);
                }
            }
        }
```

Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java]: suggestion:
I think this is more readable
```suggestion
        for (int x = -1; x <= 1; x++) {
            for (int z = -1; z <= 1; z++) {
                if (x != 0 || z != 0) {
                  res.add(pos.offset(x, 0, z));
                }
            }
        }
```

Comment (justliliandev) [reply enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java]: or:
```suggestion
        for (int x = -1; x <= 1; x++) {
            for (int z = -1; z <= 1; z++) {
                res.add(pos.offset(x, 0, z));
            }
        }
        res.remove(pos);
```

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java [hunk 5]
```diff
@@ -71,6 +112,48 @@ private boolean hasFork(ItemStack stack) {
...
+    }
+
+    private boolean removeBlock(Level level, Player player, ItemStack tool, BlockPos pos) {
+        //NB: going through all the steps rather than using level.destroyBlock so that stats are updated, hunger is used etc
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java]: BlockEvent.BreakEvent is not getting fired

File: enderio-base/src/main/java/com/enderio/base/common/lang/EIOLang.java [hunk 6]
```diff
@@ -3,18 +3,24 @@
...
 import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.TranslatableComponent;
 
 public class EIOLang {
     private static Registrate REGISTRATE = EnderIO.registrate();
 
     public static final Component COORDINATE_SELECTOR_NO_PAPER = REGISTRATE.addLang("info", EnderIO.loc("coordinate_selector.no_paper"), "No Paper in Inventory");
     public static final Component COORDINATE_SELECTOR_NO_BLOCK = REGISTRATE.addLang("info", EnderIO.loc("coordinate_selector.no_block"), "No Block in Range");
 
+    public static final TranslatableComponent ENERGY_AMOUNT = REGISTRATE.addLang("info", EnderIO.loc("energy.amount"), "%s μI");
```
Comment (justliliandev) [inline enderio-base/src/main/java/com/enderio/base/common/lang/EIOLang.java]: replace μ with \\u00B5

**Review by justliliandev — CHANGES_REQUESTED** [review]

File: enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java (hunk 2 above)
Comment (CrazyPants) [reply enderio-base/src/main/java/com/enderio/base/common/item/darksteel/DarkSteelAxeItem.java:52]: Booo! How do you see these? I tried so hard to make sure all the spaces where good! lol

**Pull request merged**